src/optimizer/plugins/AddPathRulesOptimizer.ts

```diff
@@ -1,11 +1,92 @@
 import { Package } from '../../processor';
 import { OptimizerPlugin } from '../OptimizerPlugin';
 import { ProcessingOptions } from '../../utils';
-import { ExportablePathRule } from '../../exportable';
+import {
+  ExportableAddElementRule,
+  ExportableMappingRule,
+  ExportablePathRule,
+  ExportableSdRule
+} from '../../exportable';
 import ConstructInlineInstanceOptimizer from './ConstructInlineInstanceOptimizer';
 import SimplifyArrayIndexingOptimizer from './SimplifyArrayIndexingOptimizer';
 import SimplifyRulePathContextsOptimizer from './SimplifyRulePathContextsOptimizer';
 
+// NOTE: These are the rule types that are allowed on the entities that are optimized
+// by this optimizer. Type is created to allow specificity and convenience.
+type AllowedRule =
+  | ExportableMappingRule
+  | ExportablePathRule
+  | ExportableAddElementRule
+  | ExportableSdRule;
+
+type PathNode = {
+  path: string;
+  rule?: AllowedRule[];
+  children: PathNode[];
+};
+
+// Create a tree with a node for each path part for all rules
+function buildPathTree(parent: PathNode, rule: AllowedRule) {
+  // We want to process each part, so split at each . not within []
+  // Because we want to consider each slice and index within the path as a
+  // unique path, we don't want to use the full parseFSHPath.
+  const pathParts = rule.path === '.' ? [rule.path] : rule.path.split(/\.(?![^\[]*\])/g);
+  let currentPath = '';
+  let currentPathNode: PathNode;
+  pathParts.forEach(part => {
+    currentPath = currentPath === '' ? part : `${currentPath}.${part}`; // build the path back up
+    currentPathNode = parent.children.find(child => child.path === currentPath);
+    if (currentPathNode == null) {
+      // If this path hasn't been seen, create a node for it in the tree
+      const newNode: PathNode = { path: currentPath, children: [] };
+      // If the current path is also the rule path, we want to include the rule on the node
+      if (currentPath === rule.path) {
+        newNode.rule = [rule];
+      }
+      parent.children.push(newNode);
+
+      // Use the current path's node as the parent for the next path part
+      parent = newNode;
+    } else {
+      // If the current path is also the rule path but that path's node was previously added to the tree,
+      // we still want to add the rule to node so the rule is tracked
+      if (currentPath === rule.path) {
+        currentPathNode.rule = (currentPathNode.rule ?? []).concat(rule);
+      }
+
+      // Use the current path's node as the parent for the next path part
+      parent = currentPathNode;
+    }
+  });
```

Comment:
I'm thinking a bit about the implementation inside this `forEach`. There are two main things I'm thinking about:
- The main difference between the two paths from your `if` statement is whether you have to create a new node. Apart from that, they do the same thing.
- The condition for adding `rule` to a node is true on the last iteration of the loop. Put another way, it could be done after the loop finishes, rather than having a check on every iteration.
I might be trying to get too clever here, though. Let me know if you want to talk about this.